portability/file.cc

```diff
@@ -597,20 +628,48 @@ int toku_os_close_with_source_location(
     uint src_line) {  // if EINTR, retry until success
     /* register the file close */
     int r = -1;
+    uint64_t retry_count = 0;
+    bool retry = false;
 
     /* register a file descriptor close " */
     toku_io_instrumentation io_annotation;
     toku_instr_file_fd_close_begin(
         io_annotation, toku_instr_file_op::file_close, fd, src_file, src_line);
-    while (r != 0) {
+
+    do {
         r = close(fd);
-        if (r) {
-            int rr = errno;
-            if (rr != EINTR)
-                printf("rr=%d (%s)\n", rr, strerror(rr));
-            assert(rr == EINTR);
+        if (r == -1) {
+            int what_errno = get_error_errno();
+            switch (what_errno) {
+                case ENOLCK:
+                // fallthrough
+                case EINTR:
+                // fallthrough
+                case EIO:
```

Comment:
But the issue being fixed here is that EIO must fail at once, exactly the opposite of EINTR/ENOLCK

Reply:
That is not what I came to understand about EIO, which could be incorrect. https://stackoverflow.com/questions/42434872/writing-programs-to-cope-with-i-o-errors-causing-lost-writes-on-linux

Otherwise, what is the point of this exercise at all where FT would generally assert on all errors that were not EINTR or ENOSPC.

Reply:
That's what I was referring with my comment on the TDB issue: "I didn't realise that was a release-build assert too - in which case the bug is not nearly as serious"